pager/src/main/java/dev/chrisbanes/accompanist/pager/PagerState.kt

```diff
@@ -0,0 +1,396 @@
+ * changed in any way if it has already been created.
+ *
+ * @param pageCount the initial value for [PagerState.pageCount]
+ * @param currentPage the initial value for [PagerState.currentPage]
+ * @param currentPageOffset the initial value for [PagerState.currentPageOffset]
+ */
+@ExperimentalPagerApi
+@Composable
+fun rememberPagerState(
+    pageCount: Int,
+    currentPage: Int = 0,
+    @FloatRange(from = 0.0, to = 1.0) currentPageOffset: Float = 0f,
+): PagerState = rememberSaveable(saver = PagerState.Saver) {
+    PagerState(
+        pageCount = pageCount,
+        currentPage = currentPage,
+        currentPageOffset = currentPageOffset,
+    )
+}
+
+/**
+ * A state object that can be hoisted to control and observe scrolling for [Pager].
+ *
+ * In most cases, this will be created via [rememberPagerState].
+ *
+ * @param pageCount the initial value for [PagerState.pageCount]
+ * @param currentPage the initial value for [PagerState.currentPage]
+ * @param currentPageOffset the initial value for [PagerState.currentPageOffset]
+ */
+@ExperimentalPagerApi
+class PagerState(
+    pageCount: Int,
+    currentPage: Int = 0,
+    @FloatRange(from = 0.0, to = 1.0) currentPageOffset: Float = 0f,
+) {
+    private var _pageCount by mutableStateOf(pageCount)
+    private var _currentPage by mutableStateOf(currentPage)
+    private val _currentPageOffset = mutableStateOf(currentPageOffset)
+    internal var pageSize by mutableStateOf(0)
+
+    /**
+     * The number of pages to display.
+     */
+    @get:IntRange(from = 0)
+    var pageCount: Int
+        get() = _pageCount
+        set(@IntRange(from = 0) value) {
+            _pageCount = value.coerceAtLeast(0)
+            currentPage = currentPage.coerceIn(0, lastPageIndex)
+        }
+
+    internal val lastPageIndex: Int
+        get() = (pageCount - 1).coerceAtLeast(0)
+
+    /**
+     * The index of the currently selected page.
+     *
+     * To update the scroll position, use [scrollToPage] or [animateScrollToPage].
+     */
+    @get:IntRange(from = 0)
+    var currentPage: Int
+        get() = _currentPage
+        private set(value) {
+            _currentPage = value.coerceIn(0, _pageCount)
+        }
+
+    /**
+     * The current offset from the start of [currentPage], as a fraction of the page width.
+     *
+     * To update the scroll position, use [scrollToPage] or [animateScrollToPage].
+     */
+    @get:FloatRange(from = 0.0, to = 1.0)
+    var currentPageOffset: Float
+        get() = _currentPageOffset.value
+        private set(value) {
+            _currentPageOffset.value = value.coerceIn(
+                minimumValue = 0f,
+                maximumValue = if (currentPage == lastPageIndex) 0f else 1f,
+            )
+        }
+
+    /**
+     * Represents the current selection state of a [Pager].
+     * Usually read from [PagerState.selectionState].
+     */
+    enum class SelectionState {
```

Comment:
given the new enum values is it still reasonable to call "Selection"State? but I don't think I know the better name as well

Reply:
This was going to be removed once I move to `scrollable()` anyway (since `ScrollableState` has it's own flag). I'll remove this now.